SimPEG/Regularization.py

```diff
@@ -61,320 +87,439 @@ def dim(self):
+            self._regmesh.indActive = Utils.mkvc(value)
+
+    @properties.validator('mref')
+    def _validate_mref(self, change):
+        if not isinstance(change['value'], Utils.Zero) and self.nP != '*':
+            assert len(change['value']) == self.nP, (
+                'mref must be length {}'.format(self.nP)
+            )
+
+    @properties.validator('cell_weights')
+    def _validate_cell_weights(self, change):
+        if change['value'] is not None and self.nP != '*':
+            assert len(change['value']) == self.nP, (
+                'cell_weights must be length {}'.format(self.nP)
+            )
+
+    # Other properties and methods
+    @property
+    def nP(self):
+        """
+        number of model parameters
+        """
+        if getattr(self.mapping, 'nP') != '*':
+            return self.mapping.nP
+        elif getattr(self.regmesh, 'nC') != '*':
+            return self.regmesh.nC
+        else:
+            return '*'
 
-        if mesh is not None and nP is None:
-            nP = self.regmesh.nC
-        self.nP = nP
+    @property
+    def mesh(self):
+        """
+        a SimPEG mesh which the model is described on
+        """
+        return self._mesh
 
-        self.mapping = mapping or self.mapPair(nP=self.nP)
-        self.mapping._assertMatchesPair(self.mapPair)
+    @mesh.setter
+    def mesh(self, value):
+        assert isinstance(value, Mesh.BaseMesh) or value is None, (
+            "mesh must be a SimPEG.Mesh object."
+        )
+        self._mesh = value
 
     @property
-    def parent(self):
-        """This is the parent of the regularization."""
-        return getattr(self,'_parent',None)
-    @parent.setter
-    def parent(self, p):
-        if getattr(self,'_parent',None) is not None:
-            print('Regularization has switched to a new parent!')
-        self._parent = p
+    def regmesh(self):
+        """
+        mesh used for creating operators for regularization. Excludes inactive
+        cells if they are provided
+        """
+        if getattr(self, '_regmesh', None) is None:
+            if self.indActive is not None:
+                self._regmesh = RegularizationMesh(
+                    self.mesh, indActive=self.indActive
+                )
+            else:
+                self._regmesh = RegularizationMesh(self.mesh)
+        return self._regmesh
+
+    @regmesh.setter
+    def regmesh(self, value):
+        assert isinstance(value, RegularizationMesh) or value is None, (
+            "regmesh must be an instance of a RegularizationMesh"
+            )
+        self._regmesh = value
 
     @property
-    def inv(self): return self.parent.inv
-    @property
-    def invProb(self): return self.parent
-    @property
-    def reg(self): return self
-    @property
-    def opt(self): return self.parent.opt
-    @property
-    def prob(self): return self.parent.prob
-    @property
-    def survey(self): return self.parent.survey
+    def mapping(self):
```

Comment:
`Props.Mapping` has a few additional complexities that aren't needed here (eg. the reciprocal).  I would like to give that some more thought before we bring it in